src/postgraphile/postgraphile.ts

```diff
@@ -104,11 +93,17 @@ export function getPostgraphileSchemaBuilder<
     while (true) {
       try {
         if (options.watchPg) {
-          await watchPostGraphileSchema(pgPool, pgSchemas, options, newSchema => {
-            gqlSchema = newSchema;
-            _emitter.emit('schemas:changed');
-            exportGqlSchema(gqlSchema);
-          });
+          const releaseWatchFn = await watchPostGraphileSchema(
+            pgPool,
+            pgSchemas,
+            options,
+            newSchema => {
+              gqlSchema = newSchema;
+              _emitter.emit('schemas:changed');
+              exportGqlSchema(gqlSchema);
+            },
+          );
+          shutdownActions.add(releaseWatchFn);
```

Comment:
To solve the race condition we should register a shutdown action immediately, not after the promise resolves.

```suggestion
          const releaseWatchFnPromise = watchPostGraphileSchema(
            pgPool,
            pgSchemas,
            options,
            newSchema => {
              gqlSchema = newSchema;
              _emitter.emit('schemas:changed');
              exportGqlSchema(gqlSchema);
            },
          );
          // We must register the shutdown action immediately to avoid a race condition.
          shutdownActions.add(async () => {
            const releaseWatchFn = await releaseWatchFnPromise;
            releaseWatchFn();
          });
          // Wait for the watch to be set up before progressing.
          await releaseWatchFnPromise;
```